### PR TITLE
dark magic v0.2.0: cargo

### PR DESCRIPTION
**Pull request by jhheider**
this is messy; but all other PMs will mostly work this way. so, we should get it right. but, this works at least. let's fight about better ways.

### CODE REVIEW

File: src/hooks/useExec.ts
Comment (mxcl) [inline src/hooks/useExec.ts:257]: not required surely since you defined the type as `precmd?: string`

File: src/hooks/useDarkMagic.ts
Comment (mxcl) [inline src/hooks/useDarkMagic.ts:58]: installs to `~/.local/bin/foo`, presumably we should explicitly execute that?

File: src/hooks/useExec.ts
Comment (jhheider) [reply src/hooks/useExec.ts:257]: TS was complaining. I'll take another look.

**Comment by mxcl** [comment]
I think we should provide a pkg `cargox` that then behaves like `npx` etc. in that the files are installed to a location controlled by the tool rather than how this works where the executable ends up in `~/.local/bin`

File: src/hooks/useDarkMagic.ts
Comment (jhheider) [reply src/hooks/useDarkMagic.ts:58]: We've already shown it's not on the path before installing, so that's safer but shouldn't be at issue. *Shouldn't be*...

**Comment by mxcl** [comment]
Or do it internally, but I don't think the executables should end up in a public path.

File: src/hooks/useDarkMagic.ts
Comment (mxcl) [inline src/hooks/useDarkMagic.ts:51]: Is this working via coincidence? 90% of the time I am sure the executable and pkg are named the same, but it's not guaranteed right?

Comment (jhheider) [reply src/hooks/useDarkMagic.ts:51]: Correct. We'd need some kind of massive data collection to read out the mappings.

**Comment by jhheider** [comment]
We can do that with the --root flag, so, that should be easy.

File: src/hooks/useDarkMagic.ts
Comment (mxcl) [reply src/hooks/useDarkMagic.ts:50]: massive data collection was in the spec I passed to @owenjames but that can be a next step. It's great to get something already.

I think rebuilds and glibc/linux should come first but will leave that to Owen to figure out.

File: src/hooks/usePrefix.ts
Comment (mxcl) [inline src/hooks/usePrefix.ts:12]: I'll change this after merge. I think we may go with `~/.cargo/bin` since it's the default and this works similarly to how `npx` and that work. Probs we should think about having all tea magic stuff cache inside `~/.tea` so users can clear it out. DUNNO YET!

File: tests/functional/provides.test.ts
Comment (mxcl) [inline tests/functional/provides.test.ts:45]: no longer a true comment I believe right?

**Review by mxcl — CHANGES_REQUESTED** [review]
Shouldn't we `usePrefix().bin` somewhere? Or am I missing that

File: tests/functional/provides.test.ts
Comment (jhheider) [reply tests/functional/provides.test.ts:45]: That's correct; until this moves to .cargo/bin (assuming most cargo users will add it to their path).

**Comment by jhheider** [comment]
> Shouldn't we `usePrefix().bin` somewhere? Or am I missing that

`  const binDir = usePrefix().bin`

useDarkMagic.ts:L52

**Comment by jhheider** [comment]
If it's already installed, it'll show up in the path search before getting here.

**Comment by mxcl** [comment]
> Yes, it invokes, but it doesn't do anything if it's already installed

k well that's fine. If it's quick we can leave it and not mess around.

**Review by mxcl — APPROVED** [review]

**Pull request merged**
